dmoj/judge.py

```diff
@@ -284,22 +223,238 @@ def listen(self):
+
+            exc = sys.exc_info()
+            message = ''.join(traceback.format_exception(*exc))
+
+        # Strip ANSI from the message, since this might be a checker's CompileError
+        # ...we don't want to see the raw ANSI codes from GCC/Clang on the site.
+        # We could use format_ansi and send HTML to the site, but the site doesn't presently support HTML
+        # internal error formatting.
+        self.packet_manager.internal_error_packet(strip_ansi(message))
+
+        # Logs can contain ANSI, and it'll display fine
+        print(message, file=sys.stderr)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        pass
+
+
+class SubmissionJudgeProcess:
+    def __init__(self, submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta):
+        self.submission_id = submission_id
+        self.problem_id = problem_id
+        self.language = language
+        self.source = source
+        self.time_limit = time_limit
+        self.memory_limit = memory_limit
+        self.short_circuit = short_circuit
+        self.meta = meta
+
+        self._abort_requested = False
+        self.grader = None
+        self.process = None
+        self.process_conn = None
+
+    def run(self):
+        self.process_conn, child_conn = Pipe()
+        self.process = Process(
+            name="DMOJ Judge: %s/%d Handler" % (self.problem_id, self.submission_id),
+            target=self._run_grading_wrapper_process,
+            args=(child_conn,),
+        )
+        self.process.start()
+        child_conn.close()
+
+        while True:
+            try:
+                ipc_msg = self.process_conn.recv()
+            except Exception:
+                print("Failed to read message from grading subprocess", file=sys.stderr)
+                raise
+
+            if ipc_msg is None:
+                self.process_conn.send(None)
+                break
+
+            yield ipc_msg
+
+        self.process_conn.close()
+        self.process.join()
+
+    def kill(self):
+        self.process.kill()
+
+    def join(self, timeout=None):
+        self.process.join(timeout=timeout)
+
+    def abort_grading(self):
+        if self.process:
+            try:
+                self.process_conn.send((IPC_REQUEST_ABORT, None))
+            except Exception:
+                print("Failed to send abort request, did it race?", file=sys.stderr)
+                traceback.print_exc()
+            self.join()
+
+    def _do_abort(self):
+        self._abort_requested = True
+        if self.grader:
+            self.grader.abort_grading()
+
+    # TODO(tbrindus): do we need some sort of barrier here? what happens if an abort is sent before
+    # _block_and_grade is called?
+    def _run_judge_request_handler_thread(self, conn):
+        while True:
+            try:
+                ipc_msg = conn.recv()
+            except Exception:
```

Comment:
This feels ridiculously broad.

Reply:
I don't like it either, but _so far_ I've seen this function return:

* `BrokenPipeError`
* `OSError`
* `EOFError`
* `ConnectionResetError`

I don't really want to guess at what else it can return, and none of these have a nice parent class.